molpipeline/mol2mol/filter.py

```diff
@@ -120,38 +185,318 @@ def pretransform_single(self, value: RDKitMol) -> OptionalMol:
+        """Invalidate or validate molecule matching any or all of the specified SMILES patterns.
+
+        Parameters
+        ----------
+        value: RDKitMol
+            Molecule to check.
+
+        Returns
+        -------
+        OptionalMol
+            Molecule that matches defined smiles filter, else InvalidInstance.
+        """
+        for pattern, (min_count, max_count) in self.patterns.items():
+            all_matches = value.GetSubstructMatches(Chem.MolFromSmiles(pattern))
+            if (min_count is None or len(all_matches) >= min_count) and (
+                max_count is None or len(all_matches) <= max_count
+            ):
+                if self.mode == "any":
+                    return (
+                        value
+                        if self.keep
+                        else InvalidInstance(
+                            self.uuid,
+                            f"Molecule contains forbidden SMILES pattern {pattern}.",
+                            self.name,
+                        )
+                    )
+            else:
+                if self.mode == "all":
+                    return (
+                        value
+                        if not self.keep
+                        else InvalidInstance(
+                            self.uuid,
+                            "Molecule does not match all required patterns.",
+                            self.name,
+                        )
+                    )
+        if self.mode == "any":
+            return (
+                value
+                if not self.keep
+                else InvalidInstance(
+                    self.uuid,
+                    "Molecule does not match any of the SmilesFilter patterns.",
+                    self.name,
+                )
+            )
+        return (
+            value
+            if self.keep
+            else InvalidInstance(
+                self.uuid,
+                "Molecule does not match all of the SmilesFilter patterns.",
+                self.name,
+            )
+        )
+
+
+class DescriptorsFilter(_MolToMolPipelineElement):
+    """Filter to keep or remove molecules based on RDKit descriptors."""
+
+    def __init__(
         self,
-        name: str = "MixtureFilter",
+        descriptors: dict[str, tuple[Optional[float], Optional[float]]],
+        keep: bool = True,
+        mode: Literal["any", "all"] = "any",
+        name: Optional[str] = None,
         n_jobs: int = 1,
         uuid: Optional[str] = None,
     ) -> None:
-        """Initialize MixtureFilter.
+        """Initialize DescriptorsFilter.
 
         Parameters
         ----------
-        name: str, optional (default: "MixtureFilterPipe")
+        descriptors: dict[str, tuple[Optional[float], Optional[float]]]
+            Dictionary of RDKit descriptors to filter by.
+            The value must be a tuple of minimum and maximum. If None, no limit is set.
+        keep: bool, optional (default: True)
```

Comment:
If feel we need to find a more verbose name.  Took me some time to understand what was meant.

Reply:
which one would you suggest? `keep_matches`?

Reply:
I like `keep_matches`.